src/server/pps/server/api_server.go

```diff
@@ -2362,27 +2369,96 @@ func (a *apiServer) CreatePipeline(ctx context.Context, request *pps.CreatePipel
 	// Get the expected number of workers for this pipeline
 	parallelism, err := getExpectedNumWorkers(a.env.GetKubeClient(), pipelineInfo)
 	if err != nil {
-		return nil, err
+		return err
+	}
+
+	createOrValidateSpecCommit := func() (*pfs.Commit, error) {
+		pipelineName := pipelineInfo.Pipeline.Name
+		var specCommit *pfs.Commit
+		if prevSpecCommit != nil {
+			specCommit = *prevSpecCommit
+		}
+		if specCommit != nil {
+			if !update {
+				// for new pipelines, we'll check for newness when the pipeline collection entry is created
+				return specCommit, nil
+			}
+			// make sure the pipeline branch hasn't changed outside of the transaction
+			// since the spec commit was created
+			if err := a.sudo(txnCtx.Client, func(superClient *client.APIClient) error {
```

Comment:
I'm starting to suspect some of this logic is misguided and should rely on the STM transaction protections to just work™, but I don't quite see how

Reply:
Yes, this `sudo` stuff is kind of sketchy, but for this branch in particular (when we already know the commit ID and we (probably) don't need to recreate it, I think this can be done in the transaction - that would require a `InspectCommitInTransaction`, but reads are still kind of ugly to do in transactions.